File: sql/mysql/Oracle/CSharp/MySQLLexerBase.cs
/*
 * Copyright © 2024, Oracle and/or its affiliates
 */

/* eslint-disable no-underscore-dangle */
/* cspell: ignore antlr, longlong, ULONGLONG, MAXDB */

using System;
using System.Collections.Generic;
using System.IO;
using System.Threading.Channels;
using Antlr4.Runtime;
using static System.Net.WebRequestMethods;

/** SQL modes that control parsing behavior. */
public enum SqlMode {
    NoMode,
    AnsiQuotes,
    HighNotPrecedence,
    PipesAsConcat,
    IgnoreSpace,
    NoBackslashEscapes,
};

/** The base lexer class provides a number of functions needed in actions in the lexer (grammar). */
public class MySQLLexerBase : Lexer {
    public int serverVersion = 0;
    public HashSet<SqlMode> sqlModes = new HashSet<SqlMode>();

    /** Enable Multi Language Extension support. */
    public bool supportMle = true;

    public HashSet<string> charSets = new HashSet<string>(); // Used to check repertoires.
    protected bool inVersionComment = false;

    private StackQueue<IToken> pendingTokens  = new StackQueue<IToken>();

    static string longString = "2147483647";
    static int longLength = 10;
    static string signedLongString = "-2147483648";
    static string longLongString = "9223372036854775807";
    static int longLongLength = 19;
    static string signedLongLongString = "-9223372036854775808";
    static int signedLongLongLength = 19;
    static string unsignedLongLongString = "18446744073709551615";
    static int unsignedLongLongLength = 20;

    private bool JustEmitedDot = false;

    public override string[] RuleNames => throw new NotImplementedException();

    public override IVocabulary Vocabulary => throw new NotImplementedException();

    public override string GrammarFileName => throw new NotImplementedException();


    protected MySQLLexerBase(ICharStream input, TextWriter output, TextWriter errorOutput)
	: base(input, output, errorOutput)
    {
    }

    public MySQLLexerBase(ICharStream input)
        : base(input)
    {
    }

    /**
     * Determines if the given SQL mode is currently active in the lexer.
     *
     * @param mode The mode to check.
     *
     * @returns True if the mode is one of the currently active modes.
     */
    public bool isSqlModeActive(SqlMode mode)
    {
        return this.sqlModes.Contains(mode);
    }

    /**
     * Converts a mode string into individual mode flags.
     *
     * @param modes The input string to parse.
     */
    public void sqlModeFromString(string modes)
    {
        this.sqlModes = new HashSet<SqlMode>();

        var parts = modes.ToUpper().Split(",");
        foreach (var mode in parts)
        {
            if (mode == "ANSI" || mode == "DB2" || mode == "MAXDB" || mode == "MSSQL" || mode == "ORACLE" ||
                mode == "POSTGRESQL") {
                this.sqlModes.Add(SqlMode.AnsiQuotes);
                this.sqlModes.Add(SqlMode.PipesAsConcat);
                this.sqlModes.Add(SqlMode.IgnoreSpace);
            } else if (mode == "ANSI_QUOTES") {
                this.sqlModes.Add(SqlMode.AnsiQuotes);
            } else if (mode == "PIPES_AS_CONCAT") {
                this.sqlModes.Add(SqlMode.PipesAsConcat);
            } else if (mode == "NO_BACKSLASH_ESCAPES") {
                this.sqlModes.Add(SqlMode.NoBackslashEscapes);
            } else if (mode == "IGNORE_SPACE") {
                this.sqlModes.Add(SqlMode.IgnoreSpace);
            } else if (mode == "HIGH_NOT_PRECEDENCE" || mode == "MYSQL323" || mode == "MYSQL40") {
                this.sqlModes.Add(SqlMode.HighNotPrecedence);
            }
        }
    }

    /**
     * Resets the lexer by setting initial values to transient member, resetting the input stream position etc.
     */
    public override void Reset()
    {
        this.inVersionComment = false;
        base.Reset();
    }

    /**
     * Implements the multi token feature required in our lexer.
     * A lexer rule can emit more than a single token, if needed.
     *
     * @returns The next token in the token stream.
     */
    public override IToken NextToken()
    {
        // First respond with pending tokens to the next token request, if there are any.
        var pending = this.pendingTokens.DequeueBottom();
        if (pending != null) {
            return pending;
        }

        // Let the main lexer class run the next token recognition.
        // This might create additional tokens again.
        var next = base.NextToken();

        pending = this.pendingTokens.DequeueBottom();
        if (pending != null) {
            this.pendingTokens.Push(next);

            return pending;
        }

        return next;
    }

    /**
     * Checks if the version number in the token text is less than or equal to the current server version.
     *
     * @param text The text from a matched token.
     * @returns True if so the number matches, otherwise false.
     */
    protected bool checkMySQLVersion(string text)
    {
        if (text.Length < 8) {// Minimum is: /*!12345
            return false;
        }

        // Skip version comment introducer.
        var version = Int32.Parse(text.Substring(3));
        if (version <= this.serverVersion) {
            this.inVersionComment = true;

            return true;
        }

        return false;
    }

    /**
     * Called when a keyword was consumed that represents an internal MySQL function and checks if that keyword is
     * followed by an open parenthesis. If not then it is not considered a keyword but treated like a normal identifier.
     *
     * @param proposed The token type to use if the check succeeds.
     *
     * @returns If a function call is found then return the proposed token type, otherwise just IDENTIFIER.
     */
    protected int determineFunction(int proposed)
    {
        // Skip any whitespace character if the sql mode says they should be ignored,
        // before actually trying to match the open parenthesis.
        var input = (char)this.InputStream.LA(1);
        if (this.isSqlModeActive(SqlMode.IgnoreSpace)) {
            while (input == ' ' || input == '\t' || input == '\r' || input == '\n') {
                this.Interpreter.Consume((ICharStream)this.InputStream);
                this.Channel = Hidden;
                this.Type = MySQLLexer.WHITESPACE;
                input = (char)this.InputStream.LA(1);
            }
        }

        return input == '(' ? proposed : MySQLLexer.IDENTIFIER;

    }

    /**
     * Checks the given text and determines the smallest number type from it. Code has been taken from sql_lex.cc.
     *
     * @param text The text to parse (which must be a number).
     *
     * @returns The token type for that text.
     */
    protected int determineNumericType(string text)
    {
        // The original code checks for leading +/- but actually that can never happen, neither in the
        // server parser (as a digit is used to trigger processing in the lexer) nor in our parser
        // as our rules are defined without signs. But we do it anyway for maximum compatibility.
        var length = text.Length - 1;
        if (length < MySQLLexerBase.longLength) { // quick normal case
            return MySQLLexer.INT_NUMBER;
        }

        var negative = false;
        var index = 0;
        if (text[index] == '+') { // Remove sign and pre-zeros
            ++index;
            --length;
        } else if (text[index] == '-') {
            ++index;
            --length;
            negative = true;
        }

        while (text[index] == '0' && length > 0) {
            ++index;
            --length;
        }

        if (length < MySQLLexerBase.longLength) {
            return MySQLLexer.INT_NUMBER;
        }

        int smaller;
        int bigger;
        string cmp;
        if (negative) {
            if (length == MySQLLexerBase.longLength) {
                cmp = MySQLLexerBase.signedLongString.Substring(1);
                smaller = MySQLLexer.INT_NUMBER; // If <= signed_long_str
                bigger = MySQLLexer.LONG_NUMBER; // If >= signed_long_str
            } else if (length < MySQLLexerBase.signedLongLongLength) {
                return MySQLLexer.LONG_NUMBER;
            } else if (length > MySQLLexerBase.signedLongLongLength) {
                return MySQLLexer.DECIMAL_NUMBER;
            } else {
                cmp = MySQLLexerBase.signedLongLongString.Substring(1);
                smaller = MySQLLexer.LONG_NUMBER; // If <= signed_longlong_str
                bigger = MySQLLexer.DECIMAL_NUMBER;
            }
        } else {
            if (length == MySQLLexerBase.longLength) {
                cmp = MySQLLexerBase.longString;
                smaller = MySQLLexer.INT_NUMBER;
                bigger = MySQLLexer.LONG_NUMBER;
            } else if (length < MySQLLexerBase.longLongLength) {
                return MySQLLexer.LONG_NUMBER;
            } else if (length > MySQLLexerBase.longLongLength) {
                if (length > MySQLLexerBase.unsignedLongLongLength) {
                    return MySQLLexer.DECIMAL_NUMBER;
                }
                cmp = MySQLLexerBase.unsignedLongLongString;
                smaller = MySQLLexer.ULONGLONG_NUMBER;
                bigger = MySQLLexer.DECIMAL_NUMBER;
            } else {
                cmp = MySQLLexerBase.longLongString;
                smaller = MySQLLexer.LONG_NUMBER;
                bigger = MySQLLexer.ULONGLONG_NUMBER;
            }
        }

        var otherIndex = 0;
        while (index < text.Length && cmp[otherIndex++] == text[index++]) {
            //
        }

        return text[index - 1] <= cmp[otherIndex - 1] ? smaller : bigger;
    }

    /**
     * Checks if the given text corresponds to a charset defined in the server (text is preceded by an underscore).
     *
     * @param text The text to check.
     *
     * @returns UNDERSCORE_CHARSET if so, otherwise IDENTIFIER.
     */
    protected int checkCharset(string text)
    {
        return this.charSets.Contains(text) ? MySQLLexer.UNDERSCORE_CHARSET : MySQLLexer.IDENTIFIER;
    }

    /**
     * Creates a DOT token in the token stream.
     */
    protected void emitDot()
    {
        this.pendingTokens.Push(this.TokenFactory.Create(new Tuple<ITokenSource, ICharStream>(this, (ICharStream)this.InputStream), MySQLLexer.DOT_SYMBOL,
            this.Text, this.Channel, this.TokenStartCharIndex, this.TokenStartCharIndex, this.Line,
            this.Column
        ));

        ++this.Column;
        this.JustEmitedDot = true;
    }


    // Version-related methods
    public bool isServerVersionLt80024() => serverVersion < 80024;
    public bool isServerVersionGe80024() => serverVersion >= 80024;
    public bool isServerVersionGe80011() => serverVersion >= 80011;
    public bool isServerVersionGe80013() => serverVersion >= 80013;
    public bool isServerVersionLt80014() => serverVersion < 80014;
    public bool isServerVersionGe80014() => serverVersion >= 80014;
    public bool isServerVersionGe80017() => serverVersion >= 80017;
    public bool isServerVersionGe80018() => serverVersion >= 80018;

    public bool isMasterCompressionAlgorithm() => serverVersion >= 80018 && isServerVersionLt80024();

    public bool isServerVersionLt80031() => serverVersion < 80031;

    // Functions for specific token types
    public void doLogicalOr()
    {
        Type = isSqlModeActive(SqlMode.PipesAsConcat) ? MySQLLexer.CONCAT_PIPES_SYMBOL : MySQLLexer.LOGICAL_OR_OPERATOR;
    }

    public void doIntNumber()
    {
        Type = determineNumericType(Text);
    }

    public void doAdddate() => Type = determineFunction(MySQLLexer.ADDDATE_SYMBOL);
    public void doBitAnd() => Type = determineFunction(MySQLLexer.BIT_AND_SYMBOL);
    public void doBitOr() => Type = determineFunction(MySQLLexer.BIT_OR_SYMBOL);
    public void doBitXor() => Type = determineFunction(MySQLLexer.BIT_XOR_SYMBOL);
    public void doCast() => Type = determineFunction(MySQLLexer.CAST_SYMBOL);
    public void doCount() => Type = determineFunction(MySQLLexer.COUNT_SYMBOL);
    public void doCurdate() => Type = determineFunction(MySQLLexer.CURDATE_SYMBOL);
    public void doCurrentDate() => Type = determineFunction(MySQLLexer.CURDATE_SYMBOL);
    public void doCurrentTime() => Type = determineFunction(MySQLLexer.CURTIME_SYMBOL);
    public void doCurtime() => Type = determineFunction(MySQLLexer.CURTIME_SYMBOL);
    public void doDateAdd() => Type = determineFunction(MySQLLexer.DATE_ADD_SYMBOL);
    public void doDateSub() => Type = determineFunction(MySQLLexer.DATE_SUB_SYMBOL);
    public void doExtract() => Type = determineFunction(MySQLLexer.EXTRACT_SYMBOL);
    public void doGroupConcat() => Type = determineFunction(MySQLLexer.GROUP_CONCAT_SYMBOL);
    public void doMax() => Type = determineFunction(MySQLLexer.MAX_SYMBOL);
    public void doMid() => Type = determineFunction(MySQLLexer.SUBSTRING_SYMBOL);
    public void doMin() => Type = determineFunction(MySQLLexer.MIN_SYMBOL);
    public void doNot() => Type = isSqlModeActive(SqlMode.HighNotPrecedence) ? MySQLLexer.NOT2_SYMBOL : MySQLLexer.NOT_SYMBOL;
    public void doNow() => Type = determineFunction(MySQLLexer.NOW_SYMBOL);
    public void doPosition() => Type = determineFunction(MySQLLexer.POSITION_SYMBOL);
    public void doSessionUser() => Type = determineFunction(MySQLLexer.USER_SYMBOL);
    public void doStddevSamp() => Type = determineFunction(MySQLLexer.STDDEV_SAMP_SYMBOL);
    public void doStddev() => Type = determineFunction(MySQLLexer.STD_SYMBOL);
    public void doStddevPop() => Type = determineFunction(MySQLLexer.STD_SYMBOL);
    public void doStd() => Type = determineFunction(MySQLLexer.STD_SYMBOL);
    public void doSubdate() => Type = determineFunction(MySQLLexer.SUBDATE_SYMBOL);
    public void doSubstr() => Type = determineFunction(MySQLLexer.SUBSTRING_SYMBOL);
    public void doSubstring() => Type = determineFunction(MySQLLexer.SUBSTRING_SYMBOL);
    public void doSum() => Type = determineFunction(MySQLLexer.SUM_SYMBOL);
    public void doSysdate() => Type = determineFunction(MySQLLexer.SYSDATE_SYMBOL);
    public void doSystemUser() => Type = determineFunction(MySQLLexer.USER_SYMBOL);
    public void doTrim() => Type = determineFunction(MySQLLexer.TRIM_SYMBOL);
    public void doVariance() => Type = determineFunction(MySQLLexer.VARIANCE_SYMBOL);
    public void doVarPop() => Type = determineFunction(MySQLLexer.VARIANCE_SYMBOL);
    public void doVarSamp() => Type = determineFunction(MySQLLexer.VAR_SAMP_SYMBOL);
    public void doUnderscoreCharset() => Type = checkCharset(Text);

    public bool isVersionComment() => checkMySQLVersion(Text);

    public bool isBackTickQuotedId()
    {
        return !this.isSqlModeActive(SqlMode.NoBackslashEscapes);
    }

    public bool isDoubleQuotedText()
    {
        return !this.isSqlModeActive(SqlMode.NoBackslashEscapes);
    }

    public bool isSingleQuotedText()
    {
        return !this.isSqlModeActive(SqlMode.NoBackslashEscapes);
    }

    public override IToken Emit()
    {
        IToken t = this.TokenFactory.Create(new Tuple<ITokenSource, ICharStream>(this, (ICharStream)this.InputStream),
                this.Type, (this.Text!=null?(this.JustEmitedDot?this.Text.Substring(1):this.Text):null), this.Channel, this.TokenStartCharIndex + (this.JustEmitedDot?1:0), CharIndex - 1, this.TokenStartLine, this.TokenStartColumn);
        this.JustEmitedDot = false;
        base.Emit(t);
        return t;
    }

    public void startInVersionComment()
    {
        inVersionComment = true;
    }

    public void endInVersionComment()
    {
        inVersionComment = false;
    }

    public bool isInVersionComment()
    {
        return inVersionComment;
    }
}

File: sql/mysql/Oracle/JavaScript/MySQLLexerBase.js
/*
 * Copyright © 2024, Oracle and/or its affiliates
 */
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _a, _MySQLLexerBase_longString, _MySQLLexerBase_longLength, _MySQLLexerBase_signedLongString, _MySQLLexerBase_longLongString, _MySQLLexerBase_longLongLength, _MySQLLexerBase_signedLongLongString, _MySQLLexerBase_signedLongLongLength, _MySQLLexerBase_unsignedLongLongString, _MySQLLexerBase_unsignedLongLongLength;
/* eslint-disable no-underscore-dangle */
/* cspell: ignore antlr, longlong, ULONGLONG, MAXDB */
import { Lexer, Token } from "antlr4";
import { CommonToken } from "antlr4";
import MySQLLexer from "./MySQLLexer.js";
import SqlMode from "./SqlMode.js";
import SqlModes from "./SqlModes.js";
/** The base lexer class provides a number of functions needed in actions in the lexer (grammar). */
class MySQLLexerBase extends Lexer {
    constructor(input) {
        super(input);
        this.serverVersion = 0;
        this.sqlModes = new Set();
        /** Enable Multi Language Extension support. */
        this.supportMle = true;
        this.justEmittedDot = false;
        this.charSets = new Set(); // Used to check repertoires.
        this.inVersionComment = false;
        this.pendingTokens = [];
        this.serverVersion = 80200;
        this.sqlModes = SqlModes.sqlModeFromString("ANSI_QUOTES");
    }
    /**
     * Determines if the given SQL mode is currently active in the lexer.
     *
     * @param mode The mode to check.
     *
     * @returns True if the mode is one of the currently active modes.
     */
    isSqlModeActive(mode) {
        return this.sqlModes.has(mode);
    }
    /**
     * Resets the lexer by setting initial values to transient member, resetting the input stream position etc.
     */
    reset() {
        this.inVersionComment = false;
        super.reset();
    }
    /**
     * Implements the multi token feature required in our lexer.
     * A lexer rule can emit more than a single token, if needed.
     *
     * @returns The next token in the token stream.
     */
    nextToken() {
        // First respond with pending tokens to the next token request, if there are any.
        let pending = this.pendingTokens.shift();
        if (pending) {
            return pending;
        }
        // Let the main lexer class run the next token recognition.
        // This might create additional tokens again.
        const next = super.nextToken();
        pending = this.pendingTokens.shift();
        if (pending) {
            this.pendingTokens.push(next);
            return pending;
        }
        return next;
    }
    /**
     * Checks if the version number in the token text is less than or equal to the current server version.
     *
     * @param text The text from a matched token.
     * @returns True if so the number matches, otherwise false.
     */
    checkMySQLVersion(text) {
        if (text.length < 8) { // Minimum is: /*!12345
            return false;
        }
        // Skip version comment introducer.
        const version = parseInt(text.substring(3), 10);
        if (version <= this.serverVersion) {
            this.inVersionComment = true;
            return true;
        }
        return false;
    }
    /**
     * Called when a keyword was consumed that represents an internal MySQL function and checks if that keyword is
     * followed by an open parenthesis. If not then it is not considered a keyword but treated like a normal identifier.
     *
     * @param proposed The token type to use if the check succeeds.
     *
     * @returns If a function call is found then return the proposed token type, otherwise just IDENTIFIER.
     */
    determineFunction(proposed) {
        // Skip any whitespace character if the sql mode says they should be ignored,
        // before actually trying to match the open parenthesis.
        let input = String.fromCharCode(this._input.LA(1));
        if (this.isSqlModeActive(SqlMode.IgnoreSpace)) {
            while (input === " " || input === "\t" || input === "\r" || input === "\n") {
                this._interp.consume(this._input);
                //                this.channel = Lexer.HIDDEN;
                this._type = MySQLLexer.WHITESPACE;
                input = String.fromCharCode(this._input.LA(1));
            }
        }
        return input === "(" ? proposed : MySQLLexer.IDENTIFIER;
    }
    /**
     * Checks the given text and determines the smallest number type from it. Code has been taken from sql_lex.cc.
     *
     * @param text The text to parse (which must be a number).
     *
     * @returns The token type for that text.
     */
    determineNumericType(text) {
        // The original code checks for leading +/- but actually that can never happen, neither in the
        // server parser (as a digit is used to trigger processing in the lexer) nor in our parser
        // as our rules are defined without signs. But we do it anyway for maximum compatibility.
        let length = text.length - 1;
        if (length < __classPrivateFieldGet(_a, _a, "f", _MySQLLexerBase_longLength)) { // quick normal case
            return MySQLLexer.INT_NUMBER;
        }
        let negative = false;
        let index = 0;
        if (text.charAt(index) === "+") { // Remove sign and pre-zeros
            ++index;
            --length;
        }
        else if (text.charAt(index) === "-") {
            ++index;
            --length;
            negative = true;
        }
        while (text.charAt(index) === "0" && length > 0) {
            ++index;
            --length;
        }
        if (length < __classPrivateFieldGet(_a, _a, "f", _MySQLLexerBase_longLength)) {
            return MySQLLexer.INT_NUMBER;
        }
        let smaller;
        let bigger;
        let cmp;
        if (negative) {
            if (length === __classPrivateFieldGet(_a, _a, "f", _MySQLLexerBase_longLength)) {
                cmp = __classPrivateFieldGet(_a, _a, "f", _MySQLLexerBase_signedLongString).substring(1);
                smaller = MySQLLexer.INT_NUMBER; // If <= signed_long_str
                bigger = MySQLLexer.LONG_NUMBER; // If >= signed_long_str
            }
            else if (length < __classPrivateFieldGet(_a, _a, "f", _MySQLLexerBase_signedLongLongLength)) {
                return MySQLLexer.LONG_NUMBER;
            }
            else if (length > __classPrivateFieldGet(_a, _a, "f", _MySQLLexerBase_signedLongLongLength)) {
                return MySQLLexer.DECIMAL_NUMBER;
            }
            else {
                cmp = __classPrivateFieldGet(_a, _a, "f", _MySQLLexerBase_signedLongLongString).substring(1);
                smaller = MySQLLexer.LONG_NUMBER; // If <= signed_longlong_str
                bigger = MySQLLexer.DECIMAL_NUMBER;
            }
        }
        else {
            if (length === __classPrivateFieldGet(_a, _a, "f", _MySQLLexerBase_longLength)) {
                cmp = __classPrivateFieldGet(_a, _a, "f", _MySQLLexerBase_longString);
                smaller = MySQLLexer.INT_NUMBER;
                bigger = MySQLLexer.LONG_NUMBER;
            }
            else if (length < __classPrivateFieldGet(_a, _a, "f", _MySQLLexerBase_longLongLength)) {
                return MySQLLexer.LONG_NUMBER;
            }
            else if (length > __classPrivateFieldGet(_a, _a, "f", _MySQLLexerBase_longLongLength)) {
                if (length > __classPrivateFieldGet(_a, _a, "f", _MySQLLexerBase_unsignedLongLongLength)) {
                    return MySQLLexer.DECIMAL_NUMBER;
                }
                cmp = __classPrivateFieldGet(_a, _a, "f", _MySQLLexerBase_unsignedLongLongString);
                smaller = MySQLLexer.ULONGLONG_NUMBER;
                bigger = MySQLLexer.DECIMAL_NUMBER;
            }
            else {
                cmp = __classPrivateFieldGet(_a, _a, "f", _MySQLLexerBase_longLongString);
                smaller = MySQLLexer.LONG_NUMBER;
                bigger = MySQLLexer.ULONGLONG_NUMBER;
            }
        }
        let otherIndex = 0;
        while (index < text.length && cmp.charAt(otherIndex++) === text.charAt(index++)) {
            //
        }
        return text.charAt(index - 1) <= cmp.charAt(otherIndex - 1) ? smaller : bigger;
    }
    /**
     * Checks if the given text corresponds to a charset defined in the server (text is preceded by an underscore).
     *
     * @param text The text to check.
     *
     * @returns UNDERSCORE_CHARSET if so, otherwise IDENTIFIER.
     */
    checkCharset(text) {
        return this.charSets.has(text) ? MySQLLexer.UNDERSCORE_CHARSET : MySQLLexer.IDENTIFIER;
    }
    /**
     * Creates a DOT token in the token stream.
     */
    emitDot() {
        let len = this.text.length;
        let t = new CommonToken([this, this._input], MySQLLexer.DOT_SYMBOL, 0, this._tokenStartCharIndex, this._tokenStartCharIndex);
        this.pendingTokens.push(t);
        t.text = ".";
        t.column = t.column - len;
        ++this._tokenStartCharIndex;
	this.justEmittedDot = true;
    }
    emit() {
        let t = super.emit();
	if (this.justEmittedDot) {
            t.column = t.column + 1;
            this.justEmittedDot = false;
        }
        return t;
    }
    isServerVersionLt80024() {
        return this.serverVersion < 80024;
    }
    isServerVersionGe80024() {
        return this.serverVersion >= 80024;
    }
    isServerVersionGe80011() {
        return this.serverVersion >= 80011;
    }
    isServerVersionGe80013() {
        return this.serverVersion >= 80013;
    }
    isServerVersionLt80014() {
        return this.serverVersion < 80014;
    }
    isServerVersionGe80014() {
        return this.serverVersion >= 80014;
    }
    isServerVersionGe80017() {
        return this.serverVersion >= 80017;
    }
    isServerVersionGe80018() { return this.serverVersion >= 80018; }
    isMasterCompressionAlgorithm() { return this.serverVersion >= 80018 && this.isServerVersionLt80024(); }
    isServerVersionLt80031() {
        return this.serverVersion < 80031;
    }
    doLogicalOr() {
        this._type = this.isSqlModeActive(SqlMode.PipesAsConcat) ? MySQLLexer.CONCAT_PIPES_SYMBOL : MySQLLexer.LOGICAL_OR_OPERATOR;
    }
    doIntNumber() {
        this._type = this.determineNumericType(this.text);
    }
    doAdddate() {
        this._type = this.determineFunction(MySQLLexer.ADDDATE_SYMBOL);
    }
    doBitAnd() {
        this._type = this.determineFunction(MySQLLexer.BIT_AND_SYMBOL);
    }
    doBitOr() {
        this._type = this.determineFunction(MySQLLexer.BIT_OR_SYMBOL);
    }
    doBitXor() {
        this._type = this.determineFunction(MySQLLexer.BIT_XOR_SYMBOL);
    }
    doCast() {
        this._type = this.determineFunction(MySQLLexer.CAST_SYMBOL);
    }
    doCount() {
        this._type = this.determineFunction(MySQLLexer.COUNT_SYMBOL);
    }
    doCurdate() {
        this._type = this.determineFunction(MySQLLexer.CURDATE_SYMBOL);
    }
    doCurrentDate() {
        this._type = this.determineFunction(MySQLLexer.CURDATE_SYMBOL);
    }
    doCurrentTime() {
        this._type = this.determineFunction(MySQLLexer.CURTIME_SYMBOL);
    }
    doCurtime() {
        this._type = this.determineFunction(MySQLLexer.CURTIME_SYMBOL);
    }
    doDateAdd() {
        this._type = this.determineFunction(MySQLLexer.DATE_ADD_SYMBOL);
    }
    doDateSub() {
        this._type = this.determineFunction(MySQLLexer.DATE_SUB_SYMBOL);
    }
    doExtract() {
        this._type = this.determineFunction(MySQLLexer.EXTRACT_SYMBOL);
    }
    doGroupConcat() {
        this._type = this.determineFunction(MySQLLexer.GROUP_CONCAT_SYMBOL);
    }
    doMax() {
        this._type = this.determineFunction(MySQLLexer.MAX_SYMBOL);
    }
    doMid() {
        this._type = this.determineFunction(MySQLLexer.SUBSTRING_SYMBOL);
    }
    doMin() {
        this._type = this.determineFunction(MySQLLexer.MIN_SYMBOL);
    }
    doNot() {
        this._type = this.isSqlModeActive(SqlMode.HighNotPrecedence) ? MySQLLexer.NOT2_SYMBOL : MySQLLexer.NOT_SYMBOL;
    }
    doNow() {
        this._type = this.determineFunction(MySQLLexer.NOW_SYMBOL);
    }
    doPosition() {
        this._type = this.determineFunction(MySQLLexer.POSITION_SYMBOL);
    }
    doSessionUser() {
        this._type = this.determineFunction(MySQLLexer.USER_SYMBOL);
    }
    doStddevSamp() {
        this._type = this.determineFunction(MySQLLexer.STDDEV_SAMP_SYMBOL);
    }
    doStddev() {
        this._type = this.determineFunction(MySQLLexer.STD_SYMBOL);
    }
    doStddevPop() {
        this._type = this.determineFunction(MySQLLexer.STD_SYMBOL);
    }
    doStd() {
        this._type = this.determineFunction(MySQLLexer.STD_SYMBOL);
    }
    doSubdate() {
        this._type = this.determineFunction(MySQLLexer.SUBDATE_SYMBOL);
    }
    doSubstr() {
        this._type = this.determineFunction(MySQLLexer.SUBSTRING_SYMBOL);
    }
    doSubstring() {
        this._type = this.determineFunction(MySQLLexer.SUBSTRING_SYMBOL);
    }
    doSum() {
        this._type = this.determineFunction(MySQLLexer.SUM_SYMBOL);
    }
    doSysdate() {
        this._type = this.determineFunction(MySQLLexer.SYSDATE_SYMBOL);
    }
    doSystemUser() {
        this._type = this.determineFunction(MySQLLexer.USER_SYMBOL);
    }
    doTrim() {
        this._type = this.determineFunction(MySQLLexer.TRIM_SYMBOL);
    }
    doVariance() {
        this._type = this.determineFunction(MySQLLexer.VARIANCE_SYMBOL);
    }
    doVarPop() {
        this._type = this.determineFunction(MySQLLexer.VARIANCE_SYMBOL);
    }
    doVarSamp() {
        this._type = this.determineFunction(MySQLLexer.VAR_SAMP_SYMBOL);
    }
    doUnderscoreCharset() {
        this._type = this.checkCharset(this.text);
    }
    isVersionComment() {
        return this.checkMySQLVersion(this.text);
    }
    isBackTickQuotedId() {
        return !this.isSqlModeActive(SqlMode.NoBackslashEscapes);
    }
    isDoubleQuotedText() {
        return !this.isSqlModeActive(SqlMode.NoBackslashEscapes);
    }
    isSingleQuotedText() {
        return !this.isSqlModeActive(SqlMode.NoBackslashEscapes);
    }
    startInVersionComment() {
        this.inVersionComment = true;
    }
    endInVersionComment() {
        this.inVersionComment = false;
    }
    isInVersionComment() {
        return this.inVersionComment;
    }
}
_a = MySQLLexerBase;
_MySQLLexerBase_longString = { value: "2147483647" };
_MySQLLexerBase_longLength = { value: 10 };
_MySQLLexerBase_signedLongString = { value: "-2147483648" };
_MySQLLexerBase_longLongString = { value: "9223372036854775807" };
_MySQLLexerBase_longLongLength = { value: 19 };
_MySQLLexerBase_signedLongLongString = { value: "-9223372036854775808" };
_MySQLLexerBase_signedLongLongLength = { value: 19 };
_MySQLLexerBase_unsignedLongLongString = { value: "18446744073709551615" };
_MySQLLexerBase_unsignedLongLongLength = { value: 20 };
export default MySQLLexerBase;
